proposals/0002-Extracted-Command-Line-Tools.md

```diff
@@ -0,0 +1,69 @@
+
+## Motivation
+
+Facebook does not use CLI provided with React Native, which makes it hard to maintain the project. There is no clear owner and issues / pull requests get stack in the sea of other, non-CLI related reports.
+
+It is also hard to keep moving on the CLI as every additional dependency or code change is subject to the heavy review process and sync.
+
+Having a separate package with me (and Callstack) being strategic maintainers would allow us to iterate on the design of CLI even more and make sure we have the best tooling out there.
+
+Now, if you have been following React Native development for a few years, you know I have been always super excited about CLI in general. I did many talks in that field and lead efforts of creating RNPM and then, merging it to the core as `react-native link` together with Alexey Kureev. Link, and RNPM, in general, was just the first step on our way to meet that high expectation of high-quality tools. Unfortunately, due to the aforementioned constraints and lack of time we were never able to complete our long-term strategy.
+
+I would like to take the slimmering process as an opportunity to revisit that effort and start working on it once again, together with the community.
+
+## Detailed design
+
+For the first release, the goal is to just extract out the CLI into a separate package. Since every command is a standalone module, we would like to use Yarn workspaces to build a modular architecture around it.
+
+Not many people know, but not only CLI allows you to provide your custom commands but also to override the built-in ones. That makes it even more obvious to create a package-oriented setup. The similar setup is already followed by React and Metro and we would take learnings from there.
+
+Apart from reorganizing the architecture, we would like to create a website that describes the features present in the CLI - many of them, making it much easier to work at a scale. Things like support for different locations of iOS/Android projects, linking assets from 3rd party modules are great features that only selected people were fortunate to learn about.
+
+As a next step, we would like to focus on reinvestigating every single piece of CLI to see what we can do better these days. That includes looking once again into the design of the scaffolding system (that is great, but has its limitations), lock-in to Metro (because why it should be the forced default?) and rewriting link to support modern languages and new module system. 
+
+The list of features is much longer and as a someone who believes tools are to give framework success on the market, I want to make sure we get the best tooling possible.
+
+## Drawbacks
+
+There are no drawbacks, the whole migration process happens as a part of one of the future React Native versions w/o users noticing. The separate Github repository is just for better tracking of issues.
```

Comment:
One drawback I can think of is the separation of code, and thus repositories, leading to folks not know where something lives or is implemented. While it will be good to **track** issues in two places, how will developers know which repo to file issues on? It may require some informing of developers to better understand the architecture.

Reply:
We've definitely felt this pain as a result of our CLI (https://github.com/infinitered/ignite) and boilerplate (https://github.com/infinitered/ignite-ir-boilerplate-bowser) being separated.

Reply:
(This isn't a reason _not_ to separate them, though. It's been helpful overall!)

Reply:
Yeah, I guess we should better communicate that `react-native-cli` is a separate package. I can think of an automated message printed after every failure in CLI educating that this "error report" can be submitted to this repository in particular.